### PR TITLE
[react-native] Add number[] to ActionSheetIOS destructiveButtonIndex

### LINKED ISSUE #58258: [react-native]: Add number[] to ActionSheetIOS destructiveButtonIndex
### Discussed in https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/58256

<div type='discussions-op-text'>

<sup>Originally posted by **rcb4t2** January 17, 2022</sup>
In ActionSheetIOS, the `destructiveButtonIndex` is typed as a number

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/97ee04ae27d5850c7775701e2d0634f44b3ee971/types/react-native/v0.63/index.d.ts#L6905

But it can also accept an array of numbers

https://reactnative.dev/docs/actionsheetios#showactionsheetwithoptions

Please update the definition - we are using v63 so we would appreciate updating that version as well as 64 and 65

Thank you!</div>

### PR DESCRIPTION
**Pull request by rcb4t2**
Closes #58258

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [full details](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/58256)

### CODE REVIEW

**Review by ca057 — COMMENTED** [review]
We should also update it in the [main definition file](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/13e19e082e3fde946f16b8185d9fd18d63277a24/types/react-native/index.d.ts) respectively. Can we add some tests for this change?

In the documentation for v0.63 (https://reactnative.dev/docs/0.63/actionsheetios) I can’t see that either `cancelButtonIndex` nor `destructiveButtonIndex` is potentially an array of numbers. Do you have a source? The linked documentation points to the current version.

**Comment by rcb4t2** [comment]
@ca057 Thanks Cristian! I agree that the 0.63 docs do not show a number array, but it looks like an error in the docs
The source code on that branch shows that number array is accepted, and I can confirm that an array works as expected when given in the props
https://github.com/facebook/react-native/blob/4f897336b6ce6c640a65f86b65c6d386ea55b944/Libraries/ActionSheetIOS/ActionSheetIOS.js#L47

Lmk what you think!

I'll go ahead and update the main definition file as well. Can you help me understand what kind of tests to add? I'm not familiar with this kind of testing framework. Do I just call the ActionSheetIOS.showActionSheetWithOptions in the test file?

Thanks!

**Comment by rcb4t2** [comment]
@ca057 Please see a proposed passing test here: 
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58257/commits/7922a589ca286ce0afe240e0988b7e85f3e64371

If that looks good to you I'll add it into the v0.63-0.65 folders too

File: types/react-native/v0.63/index.d.ts [hunk 1]
```diff
@@ -6901,7 +6901,7 @@ export class SwipeableListView extends React.Component<SwipeableListViewProps> {
 export interface ActionSheetIOSOptions {
     title?: string;
     options: string[];
-    cancelButtonIndex?: number;
+    cancelButtonIndex?: number | number[];
     destructiveButtonIndex?: number;
```
Comment (ca057) [inline types/react-native/v0.63/index.d.ts:6905]: I think this is a mistake here and should be the other way, right?

```suggestion
    cancelButtonIndex?: number;
    destructiveButtonIndex?: number | number[] | undefined | null;
```

File: types/react-native/v0.64/index.d.ts [hunk 2]
```diff
@@ -6930,7 +6930,7 @@ export interface ActionSheetIOSOptions {
     title?: string | undefined;
     options: string[];
     cancelButtonIndex?: number | undefined;
-    destructiveButtonIndex?: number | undefined;
+    destructiveButtonIndex?: number | number[] | undefined;
```
Comment (ca057) [inline types/react-native/v0.64/index.d.ts:6933]: We could improve it here, as Maybe types (see the original source code written in Flow) would also accept `null`: https://flow.org/en/docs/types/maybe/
```suggestion
    destructiveButtonIndex?: number | number[] | undefined | null;
```

File: types/react-native/v0.65/index.d.ts [hunk 3]
```diff
@@ -7007,7 +7007,7 @@ export interface ActionSheetIOSOptions {
     title?: string | undefined;
     options: string[];
     cancelButtonIndex?: number | undefined;
-    destructiveButtonIndex?: number | undefined;
+    destructiveButtonIndex?: number | number[] | undefined;
```
Comment (ca057) [inline types/react-native/v0.65/index.d.ts:7010]: Same here :)

**Comment by ca057** [comment]
> @ca057 Thanks Cristian! I agree that the 0.63 docs do not show a number array, but it looks like an error in the docs. The source code on 0.63 - 0.65 shows that number array is accepted, and I can confirm that an array works as expected when given in the props https://github.com/facebook/react-native/blob/v0.63.4/Libraries/ActionSheetIOS/ActionSheetIOS.js#L47 https://github.com/facebook/react-native/blob/v0.64.3/Libraries/ActionSheetIOS/ActionSheetIOS.js#L48 https://github.com/facebook/react-native/blob/v0.65.2/Libraries/ActionSheetIOS/ActionSheetIOS.js#L46
> 
> Lmk what you think!
> 
> I'll go ahead and update the main definition file as well. Can you help me understand what kind of tests to add? I'm not familiar with this kind of testing framework. Do I just call the ActionSheetIOS.showActionSheetWithOptions in the test file?
> 
> Thanks!

Thanks for digging out the different sources, I was more wondering about the `cancelButtonIndex` you once made an array, I guess this was an error. The tests looks good for me as well 👌 

Seems good for me to go once we clarified the question around the `cancelButtonIndex`.

**Comment by rcb4t2** [comment]
@ca057 Thanks Christian! You are right, the `cancelButtonIndex` was an error. I will remove that, incorporate your suggestions re: `null`, and add coverage to each version

**Comment by rcb4t2** [comment]
Hi @ca057! This PR is all set for review, but it appears there is a failing test unrelated to my changes in `styled-components-react-native`. Please advise how I should proceed, thanks!

**Review by eps1lon — APPROVED** [review]

**Comment by rcb4t2** [comment]
@eps1lon @ca057 I'm not sure what's going on with the `styled-components-react-native` test. I tried merging from master again today b/c I saw a change related to that package, but no luck. I will stop altering this PR while I await further instructions

**Comment by rcb4t2** [comment]
@eps1lon Can you take a look at this? It looks like either a CI bug or a problem in the `styled-components-react-native` types file

**Comment by eps1lon** [comment]
@rcb4t2 Try rebasing this PR with `master` maybe this'll fix CI issues.

**Comment by rcb4t2** [comment]
> @rcb4t2 Try rebasing this PR with `master` maybe this'll fix CI issues.

@eps1lon Thanks! First I tried a simple merge from master, then a rebase and force push. Same CI problem with both. I don't understand what's going on here, but I don't think it has anything to do with the changes I made

I don't know how much more time I'm willing to put towards this. Lmk if you have any suggestions, or maybe I'll just leave this change for someone else and override the type definition locally in my repo

**Comment by rcb4t2** [comment]
@eps1lon Update on this: if I explicitly `npm install` the styled-components library deps, the tests pass on my local environment. It seems like the CI skips installing those deps. I don't know how to change that behavior. Any thoughts?

I also created a new branch from the latest master and added my changes, and the tests still fail at styled-components. Seems pretty clear this is a bug in the CI. How do you want to proceed?

**Comment by sandersn** [comment]
I tried pulling the branch and running the CI tests locally and it passed. I'm going to merge with master and push to see whether CI agrees with my local machine.

**Comment by sandersn** [comment]
Oh, it's very likely because I ran `npm install` on `styled-components` as part of testing the overnight test run locally, and CI's incremental nature doesn't do this, and misses the dependency from styled-components-react-native to styled-components.

This is a styled-components-react-native problem with dependencies, not a react-native one, and it's *probably intentional* to avoid transitive dependencies on both the browser types and the node types at the same time. I think we can merge this despite the CI failure.

**Comment by sandersn** [comment]
Filed https://github.com/microsoft/DefinitelyTyped-tools/issues/413 to track this.

**Comment by rcb4t2** [comment]
@sandersn Awesome Nathan! Thanks for chasing this down!

**Pull request merged**